File: achieveit_common/src/main/java/com/april/achieveit_common/utility/SnowFlakeIdGenerator.java
package com.april.achieveit_common.utility;

import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.lang3.RandomStringUtils;
import org.apache.commons.lang3.RandomUtils;
import org.apache.commons.lang3.StringUtils;

public class SnowFlakeIdGenerator
{
    //Initial Timestamp
    private static final long START_TIMESTAMP=1514736000000L;
    private static final int SERIAL_BITS_LENGTH=14;
    private static final int MACHINE_BITS_LENGTH=6;
    private static final int DATA_CENTER_BITS_LENGTH=2;

    private static final long MAXIMUM_SERIAL_NUM=~(-1L<<SERIAL_BITS_LENGTH);
    private static final long MAXIMUM_MACHINE_NUM=~(-1L<<MACHINE_BITS_LENGTH);
    private static final long MAXIMUM_DATA_CENTER_NUM=~(-1L<<DATA_CENTER_BITS_LENGTH);

    private static final int SERIAL_OFFSET=0;
    private static final int MACHINE_OFFSET=SERIAL_OFFSET+SERIAL_BITS_LENGTH;
    private static final int DATA_CENTER_OFFSET=MACHINE_OFFSET+MACHINE_BITS_LENGTH;
    private static final int TIMESTAMP_OFFSET=DATA_CENTER_OFFSET+DATA_CENTER_BITS_LENGTH;

    private long dataCenterId;
    private long machineId;
    private long serial=0L;
    private long lastTimeStamp=-1L;

    public SnowFlakeIdGenerator(long dataCenterId,long machineId)
    {
        if(dataCenterId>MAXIMUM_DATA_CENTER_NUM||dataCenterId<0)
        {
            throw new IllegalArgumentException("Invalid data center id.");
        }
        if(machineId>MAXIMUM_MACHINE_NUM||machineId<0)
        {
            throw new IllegalArgumentException("Invalid machine id.");
        }
        this.dataCenterId=dataCenterId;
        this.machineId=machineId;
        lastTimeStamp=System.currentTimeMillis();
    }

    public synchronized long getNextId()
    {
        long currentStamp=System.currentTimeMillis();
        if(currentStamp<lastTimeStamp)
        {
            throw new RuntimeException("Clock move backwards. Quitting due to unacceptable fatal error.");
        }
        if(currentStamp==lastTimeStamp)
        {
            serial=(serial+1)&MAXIMUM_SERIAL_NUM;
            if(serial==0L)
            {
                waitUntilNextMillisecond();
                currentStamp=System.currentTimeMillis();
            }
        }
        else
        {
            serial=0L;
        }
        lastTimeStamp=currentStamp;
        long result=(currentStamp-START_TIMESTAMP)<<TIMESTAMP_OFFSET|dataCenterId<<DATA_CENTER_OFFSET|machineId<<MACHINE_OFFSET|serial<<SERIAL_OFFSET;
        return result;
    }

    private void waitUntilNextMillisecond()
    {
        long currentStamp=System.currentTimeMillis();
        for(;currentStamp<=lastTimeStamp;)
            currentStamp=System.currentTimeMillis();
    }
}

File: Artifact/snowflake_id_generator.py
from datetime import datetime


class SnowflakeIdGenerator(object):
    START_TIMESTAMP: int = int(datetime(2018, 1, 1, 0, 0, 0).timestamp() * 1000)
    SERIAL_BITS_LENGTH: int = 14
    MACHINE_BITS_LENGTH: int = 6
    DATA_CENTER_BITS_LENGTH: int = 2

    MAXIMUM_SERIAL_NUM = ~(-1 << SERIAL_BITS_LENGTH)
    MAXIMUM_MACHINE_NUM = ~(-1 << MACHINE_BITS_LENGTH)
    MAXIMUM_DATA_CENTER_NUM = ~(-1 << DATA_CENTER_BITS_LENGTH)

    SERIAL_OFFSET: int = 0
    MACHINE_OFFSET: int = SERIAL_OFFSET + SERIAL_BITS_LENGTH
    DATA_CENTER_OFFSET: int = MACHINE_OFFSET + MACHINE_BITS_LENGTH
    TIMESTAMP_OFFSET: int = DATA_CENTER_OFFSET + DATA_CENTER_BITS_LENGTH

    def __init__(self, data_center_id: int, machine_id: int):
        self.data_center_id: int = data_center_id
        self.machine_id: int = machine_id
        self.serial: int = 0
        self.last_time_stamp: int = int(datetime.now().timestamp() * 1000)

    def get_next_id(self) -> int:
        current_stamp: int = int(datetime.now().timestamp() * 1000)
        if current_stamp < self.last_time_stamp:
            raise RuntimeError("Clock move backwards. Quitting due to unacceptable fatal error.")

        if current_stamp == self.last_time_stamp:
            self.serial = (self.serial + 1) & SnowflakeIdGenerator.MAXIMUM_SERIAL_NUM
            if self.serial == 0:
                self.wait_until_next_millisecond()
                current_stamp = int(datetime.now().timestamp() * 1000)
        else:
            self.serial = 0
        self.last_time_stamp = current_stamp
        result: int = int(current_stamp - SnowflakeIdGenerator.START_TIMESTAMP) \
                      << SnowflakeIdGenerator.TIMESTAMP_OFFSET \
                      | self.data_center_id << SnowflakeIdGenerator.DATA_CENTER_OFFSET \
                      | self.machine_id << SnowflakeIdGenerator.MACHINE_OFFSET \
                      | self.serial << SnowflakeIdGenerator.SERIAL_OFFSET
        return result

    def wait_until_next_millisecond(self):
        current_stamp = datetime.now().timestamp() * 1000
        while current_stamp <= self.last_time_stamp:
            current_stamp = datetime.now()


if __name__ == '__main__':
    id_generator = SnowflakeIdGenerator(machine_id=0, data_center_id=0)
    for i in range(1, 200):
        print(id_generator.get_next_id())
